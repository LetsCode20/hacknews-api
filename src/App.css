* {
  padding: 0;
  margin: 0;
}

html {
  box-sizing: border-box;
}

body {
  color: #222;
}