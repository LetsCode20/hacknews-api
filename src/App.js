import React, { useEffect, useState } from 'react';
import Search from './components/Search.component';
import Table from './components/Table.component';

const App = () => {
  const listObj = [
    {
      title: 'React',
      url: 'https://facebook.github.io/react/',
      author: 'Jordan Walke',
      num_comments: 3,
      points: 4,
      objectID: 0,
    },
    {
      title: 'Redux',
      url: 'https://github.com/reactjs/redux',
      author: 'Dan Abramov, Andrew Clark',
      num_comments: 2,
      points: 5,
      objectID: 1,
    },
  ];
  const DEFAULT_QUERY = 'redux';
  const PATH_BASE = 'https://hn.algolia.com/api/v1';
  const PATH_SEARCH = '/search';
  const PARAM_SEARCH = 'query=';
  const PATH_URL = `${PATH_BASE}${PATH_SEARCH}${PARAM_SEARCH}${DEFAULT_QUERY}`;

  const [list, setList] = useState(listObj);
  const [searchTerm, setSearchTerm] = useState(DEFAULT_QUERY);
  const [result, setResult] = useState(null);

  const setSearchTopStories = (result) => {
    setResult({ result });
  };

  useEffect(() => {
    fetch(`${PATH_BASE}${PATH_SEARCH}${PARAM_SEARCH}${searchTerm}`)
      .then((response) => response.json())
      .then((result) => setSearchTopStories(result))
      .catch((error) => error);
  }, [searchTerm]);

  const onSearchChange = (e) => {
    setSearchTerm(e.target.value);
  };

  const onDelete = (id) => {
    const isNotId = (item) => item.objectID !== id;

    const updateList = list.filter(isNotId);
    setList(updateList);
  };

  const isSearched = (searchTerm) => (item) =>
    item.title.toLowerCase().includes(searchTerm.toLowerCase());

  if (!result) {
    return null;
  }

  return (
    <div className='page'>
      <div className='interactions'>
        <Search value={searchTerm} onChange={onSearchChange} />
      </div>

      <Table
        list={result.hits}
        pattern={searchTerm}
        onDelete={onDelete}
        isSearched={isSearched}
      />
    </div>
  );
};

export default App;
