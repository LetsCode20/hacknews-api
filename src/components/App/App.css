@import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@500;600;700&display=swap');

* {
  padding: 0;
  margin: 0;
}

*:focus {
  outline: none;
}

html {
  box-sizing: border-box;
  font-family: 'Montserrat', sans-serif;
}

body {
  color: #222;
  background: #f4f4f4;
}

a {
  color: #222;
  text-decoration: none;
}

a:hover {
  text-decoration: underline;
}

input {
  padding: 10px;
  border-radius: 5px;
  outline: none;
  margin-right: 10px;
  border: 1px solid #dddddd;
}

button {
  padding: 10px;
  margin: 10px;
  border-radius: 5px;
  border: 1px solid #dddddd;
  background: transparent;
  color: #808080;
  cursor: pointer;
}

button:hover {
  color: #222;
}

.page {
  margin: 20px;
}

.interactions {
  text-align: center;
}

.sort {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
}

.sort button {
  background-color: #222;
  color: #fff;
}

.sort button:hover {
  background-color: #333;
  color: #fff;
}

.table-row {
  width: 90%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  line-height: 24px;
  margin: 10px auto;
  padding: 10px;
  background: #ffffff;
  border: 1px solid #e3e3e3;
}

.table-column {
  width: 20%;
}

.button-inline {
  color: #fff;
  background: #e10;
}

.button-inline:hover {
  color: #fff;
  background: #f00;
}

@media (max-width: 991px) {
  .table-row {
    flex-direction: column;
    justify-content: center;
  }

  .table-column {
    width: 100%;
    text-align: center;
  }

  .table-column span {
    line-height: 2;
  }

  .table-column b {
    margin-right: 10px;
  }
}
