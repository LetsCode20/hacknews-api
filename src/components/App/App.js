import React, { useEffect, useState } from 'react';
import axios from 'axios';

import Button from '../Button/Button.component';
import Search from '../Search/Search.component';
import Table from '../Table/Table.component';
import {
  DEFAULT_QUERY,
  DEFAULT_HPP,
  PATH_BASE,
  PATH_SEARCH,
  PARAM_SEARCH,
  PARAM_PAGE,
  PARAM_HPP,
} from '../../constants/constants';

const App = () => {
  const [searchTerm, setSearchTerm] = useState(DEFAULT_QUERY);
  const [results, setResults] = useState(null);
  const [searchKey, setSearchKey] = useState('');
  const [error, setError] = useState(null);

  const page = (results && results[searchKey] && results[searchKey].page) || 0;
  const list = (results && results[searchKey] && results[searchKey].hits) || [];

  const onSearchChange = (e) => {
    setSearchTerm(e.target.value);
  };

  const onDelete = (id) => {
    const { hits, page } = results[searchKey];

    const isNotId = (item) => item.objectID !== id;
    const updatehits = hits.filter(isNotId);
    const updateResult = {
      ...results,
      [searchKey]: { hits: updatehits, page },
    };

    setResults(updateResult);
  };

  const needToSearchTopStories = (searchTerm) => !results[searchTerm];

  const setSearchTopStories = (result) => {
    const { hits, page } = result;

    const oldHits = results && results[searchKey] ? result[searchKey].hits : [];
    const updatehits = [...oldHits, ...hits];

    setResults({ ...results, [searchKey]: { hits: updatehits, page } });
  };

  const fetchSearchTopStories = (searchTerm, page = 0) => {
    axios(
      `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
    )
      .then((result) => setSearchTopStories(result.data))
      .catch((error) => setError(error));
  };

  const onSearchSubmit = (event) => {
    event.preventDefault();
    setSearchKey(searchTerm);

    if (needToSearchTopStories(searchTerm)) {
      fetchSearchTopStories(searchTerm);
    }
  };

  useEffect(() => {
    fetchSearchTopStories(searchTerm);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchTerm]);

  return (
    <div className='page'>
      <div className='interactions'>
        <Search
          value={searchTerm}
          onChange={onSearchChange}
          onSubmit={onSearchSubmit}
        >
          Search
        </Search>
      </div>
      {error ? (
        <div className='interactions'>
          <p>Something went wrong.</p>
        </div>
      ) : (
        results && <Table list={list} onDelete={onDelete} />
      )}
      <div className='interactions'>
        <Button onClick={() => fetchSearchTopStories(searchTerm, page + 1)}>
          Click to See More
        </Button>
      </div>
    </div>
  );
};

export default App;
