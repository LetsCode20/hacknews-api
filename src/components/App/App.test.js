import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from './App.js';

describe('App', () => {
  it('renders without errors or crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
