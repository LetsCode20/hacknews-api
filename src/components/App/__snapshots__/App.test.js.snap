// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button>
        Search
      </button>
    </form>
  </div>
  <div
    className="interactions"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Click to See More
    </button>
  </div>
</div>
`;
